### PR TITLE
fix: let unary traits work at comptime

### LINKED ISSUE #5492: Let `!` and `-` operators rely on traits at comptime
### Aim

In #5423 we made these operators work at runtime, but they still don't work at comptime.

### Expected Behavior

This should compile fine:

```rust
struct MyType {}

impl Neg for MyType { ... }

fn main() {
    comptime {
        let value = MyType {};
        let _result = -value;
    }
}
```

### Bug

It doesn't compile.

### To Reproduce

_No response_

### Project Impact

None

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Installation Method

None

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #5492

## Summary

In #5423 we made unary traits work at runtime: this PR makes them work at comptime.

## Additional Context

For some reason, in the test program, `_result` ends up being `()` when I print it. It's the same if I implement `Add` for a type and use it a comptime, so maybe it's expected or maybe that's not yet implemented?

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
> For some reason, in the test program, _result ends up being () when I print it. It's the same if I implement Add for a type and use it a comptime, so maybe it's expected or maybe that's not yet implemented?

This is because the comptime block being run is elaborated before your impl method. Your method is comptime but the impl as a whole is not and the compiler didn't add it to the list of comptime items so it just sees an empty function body. It's an issue I'm tackling in another branch right now so I imagine this will need to stay up until I can think of a solution for it.

For now, can you add an assert to the test that the negative value is as expected? It won't work right now due to the unit problem you mentioned but I'd like to have it for the future.

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs [hunk 1]
```diff
@@ -593,7 +593,12 @@ impl<'a> Interpreter<'a> {
 
     fn evaluate_prefix(&mut self, prefix: HirPrefixExpression, id: ExprId) -> IResult<Value> {
         let rhs = self.evaluate(prefix.rhs)?;
-        self.evaluate_prefix_with_value(rhs, prefix.operator, id)
+
+        if self.interner.get_selected_impl_for_expression(id).is_some() {
+            self.evaluate_overloaded_prefix(prefix, rhs, id);
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/comptime/interpreter.rs:598]:
```suggestion
            self.evaluate_overloaded_prefix(prefix, rhs, id)
```

**Comment by asterite** [comment]
> For now, can you add an assert to the test that the negative value is as expected? It won't work right now due to the unit problem you mentioned but I'd like to have it for the future.

I'm not sure how. If I do this:

```rust
    comptime
    {
        let value = MyType { value: 1 };
        let result = -value;
        assert_eq(result.value, -1);
    }
```

I get this error:

```
error: The type `()` has no fields to access
   ┌─ src/main.nr:34:19
   │
34 │         assert_eq(result.value, -1);
   │                   ------------
```

**Comment by jfecher** [comment]
@asterite That's fine - it's what I meant about it not working now.
We may have to wait until this bug is fixed to test this PR anyway.

**Review by jfecher — APPROVED** [review]
:+1:

**Review by michaeljklein — APPROVED** [review]
LGTM

**Comment by asterite** [comment]
@jfecher Now this program doesn't compile anymore, do you know what could have happened? (or maybe it's expected?)

```rust
struct MyType {
    value: i32,
}

impl Neg for MyType {
    comptime fn neg(self) -> Self {
        self
    }
}

fn neg_at_comptime() {
    comptime
    {
        let value = MyType { value: 1 };
        let _result = -value;
    }
}
```

It says "`neg` cannot be called in a `comptime` context here". And it seems to come from this change here: https://github.com/noir-lang/noir/commit/6a7f593a04ee1caefd6a19a5cba1c0dbeee22ee1#diff-3475af38ea5c8491a0a0dc1c32ac9f194c52fda955c8cb577582e94a67e15464R132-R135

**Comment by jfecher** [comment]
@asterite that check fails for any functions whose bodies are still empty at that point

**Comment by TomAFrench** [comment]
It does look like that program should probably display a warning similar to how it does for `pub` however as it's ignoring the `comptime`.

**Comment by asterite** [comment]
> For now you should make the impl itself comptime

Thanks, that worked!
